### ISSUE TITLE
GetMyIP fails on iOS/OSX devices

### ISSUE THREAD

**Issue by ArnisLielturks** [SLikeSoft/SLikeNet#36]
**GetMyIP_Windows_Linux_IPV4()**  function is unable to get local IP address by using the hostname.
It seems that by default iOS/OSX systems doesn't contain  hostname record in the /etc/hosts file. I believe the same issue would appear by using the IPV6 support (**GetMyIP_Windows_Linux_IPV4And6**).

The proposed fix to this:

```
void GetMyIP_Windows_Linux_IPV4( SystemAddress addresses[MAXIMUM_NUMBER_OF_INTERNAL_IDS] )
{



	int idx=0;
	char ac[ 80 ];
	int err = gethostname( ac, sizeof( ac ) );
        (void) err;
	RakAssert(err != -1);
    const char *localhost_str="localhost"; // <------
	
	struct addrinfo *curAddress = NULL;
	err = getaddrinfo(ac, NULL, NULL, &curAddress);

    if(curAddress == NULL) // <------
    { // <------
        err = getaddrinfo(localhost_str, NULL, NULL, &curAddress); // <------
    } // <------
    
	if ( err != 0 || curAddress == 0 )
	{
		RakAssert(false);
		return ;
	}
	while (curAddress != NULL && idx < MAXIMUM_NUMBER_OF_INTERNAL_IDS)
	{
		if (curAddress->ai_family == AF_INET) {
			addresses[idx].address.addr4 = *((struct sockaddr_in *)curAddress->ai_addr);
			++idx;
		}
		curAddress = curAddress->ai_next;
	}
	
	while (idx < MAXIMUM_NUMBER_OF_INTERNAL_IDS)
	{
		addresses[idx]=UNASSIGNED_SYSTEM_ADDRESS;
		idx++;
	}

}
```

If the hostname record is added manually to the /etc/hosts file, everything works as expected.

More information about the issue is available here, where SLikeNet is used as a 3rd party library: https://github.com/urho3d/Urho3D/pull/2302

**Comment by Luke1410** [comment]
Thanks for the report and suggested fix. Assigned internal case number: SLNET-217 (scheduled for the upcoming SLikeNet 0.2.0 release).

In order to apply the patch in SLikeNet, we'd like to ask you agree with our CLA: https://github.com/SLikeSoft/SLikeNet/blob/master/.github/CONTRIBUTING.md
and let us know if you'd like to be given credit for the contribution in the readme's acknowledgment section.

I find it hard atm to get any concrete first hand information explaining the underlying behavior of gethostname()/getaddrinfo() on OSX/iOS. The closest information/explanation I could find is in this stackoverflow post: https://stackoverflow.com/questions/20333364/got-local-host-name-working-on-windows-but-not-in-linux

Can't see any relevant docu on the developer.apple.com page either: https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man3/gethostname.3.html

Do you have any further resource which would explain the behavior you see on OSX/iOS?
Which string does gethostname() return for you?
If having the fix applied to the current Git master branch of SLikeNet, pls let me know and I'll integrate it directly.

**Comment by ArnisLielturks** [comment]
@Luke1410 The problem is that on some workstations the /etc/hosts file doesn't contain record which assigns hostname to 127.0.0.1 and ::1 addresses. If the record is added manually, everything works as expected. Of course this covers only the OSX devices, iOS devices still have the same problem since the /etc/hosts file is not available for editing.

Additional details are explained in the link provided in the first comment (comments related to the problem are at the very bottom of the PR). 

I agree with your CLA and you don't have to add me to the contributors list.

**Comment by Luke1410** [comment]
Thanks for accepting the CLA. I put you on record now.

The issue and behavior of gethostname()/getaddrinfo(), I understand of course. What I didn't yet get is the exact rational and details why Apple decided to go a different way.

Digging into the behavior on OSX a bit further I'm getting a better picture now on the details behind that  --- we'll fix this properly by rewriting the related functions which should do it in a more appropriate/proper way.

Let me know if having your temporary workaround in the SLikeNet master repository available meanwhile would be of help to you. Otherwise I'll get back to this issue once we have the new approach in place (due to other priorities atm, this might take a couple of days though).

**Comment by ArnisLielturks** [comment]
@Luke1410 That's fine if the fix can be implemented in the next week or something, it's not a critical issue since we know the workaround to it. You can leave out the temporary fix from the master branch since you have an idea how to fix that properly. 

Also thanks for the quick responses!

**Comment by ArnisLielturks** [comment]
Hi!

Any updates on this?

**Comment by Luke1410** [comment]
We are working atm on some outstanding security reports. Afterwards getting this issue resolved is next on our list. As said, if it's urgent we can apply the temporary solution you suggested; otherwise rest assured that it's still the top priority issue on our list directly after having dealt with the security related tasks.

**Comment by ArnisLielturks** [comment]
@Luke1410 No,  that's fine. Just wanted to know the status of this, that's all.

**Comment by Luke1410** [comment]
Just a quick status update:
We completed the security issue related work yesterday and started looking into this issue now. We just completed the rewrite of determining the IP address for the own machine for Windows (committed to the SVN repository) and are working on the Linux/OSX part next.

**Comment by Luke1410** [comment]
The issue should have been resolved. Changes are available on master in the Git repository and on trunk in the SVN repository.
If you rely on the changes/fixes being made available in a released version soon, please let us know. Otherwise the changes are expected to be shipped with SLikeNet 0.2.0.

Note that we only partially tested the fix for Linux/OSX (a complete testrun will be done as part of the 0.2.0 release process). If you are able to test whether your issue was resolved, we'd appreciate the feedback.

**Comment by ArnisLielturks** [comment]
@Luke1410 Thanks for this! Will get the latest version from the master branch and test it out!
